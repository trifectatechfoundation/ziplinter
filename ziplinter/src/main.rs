use rc_zip_sync::ReadZip;
use std::ops::Deref;

fn main() {
    let mut args = std::env::args();
    let _ = args.next();
    let path = args.next().unwrap();

    let file = std::fs::File::open(path).unwrap();
    let archive = file.read_zip().unwrap();

    eprintln!(
        "{}",
        serde_json::to_string_pretty(&archive.deref()).unwrap()
    )
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::assert_json_snapshot;
    use std::{error::Error, path::Path};

    fn process_zip_file(zip_path: &Path) -> Result<serde_json::Value, Box<dyn Error>> {
        let file = std::fs::File::open(zip_path).unwrap();
        let archive = file.read_zip();

        Ok(serde_json::to_value(archive?.deref())?)
    }

    #[test]
    fn snapshot_zip_files() {
        let fixtures_dir = std::env::current_dir()
            .unwrap()
            .join("../testdata")
            .canonicalize()
            .unwrap();

        println!("fixtures_dir: {}", fixtures_dir.display());

        for entry in std::fs::read_dir(fixtures_dir).expect("Failed to read fixtures directory") {
            let entry = entry.expect("Failed to read entry");
            let path = entry.path();

            if path
                .extension()
                .is_some_and(|e| e.eq_ignore_ascii_case("zip"))
            {
                let Ok(result) = process_zip_file(&path) else {
                    continue;
                };
                println!(
                    "current file: {}",
                    Path::new(path.file_name().unwrap()).display()
                );
                assert_json_snapshot!(
                    format!("{}", Path::new(path.file_name().unwrap()).display()),
                    result
                );
            }
        }
    }
}
