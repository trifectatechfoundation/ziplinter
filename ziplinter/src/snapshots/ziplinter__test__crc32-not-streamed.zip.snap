---
source: ziplinter/src/main.rs
expression: result
---
{
  "comment": "",
  "contents": [
    {
      "central": {
        "comment": "",
        "compressed_size": 4,
        "crc32": 2117232040,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          222,
          85,
          89,
          79,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 0,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 420,
        "modified": "2012-03-09T00:59:10Z",
        "name": "foo.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 4
      },
      "local": {
        "accessed": null,
        "compressed_size": 4,
        "crc32": 2117232040,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          222,
          85,
          89,
          79,
          222,
          85,
          89,
          79,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "gid": 501,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2012-03-09T00:59:10Z",
        "name": "foo.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 501,
        "uncompressed_size": 4
      }
    },
    {
      "central": {
        "comment": "",
        "compressed_size": 4,
        "crc32": 77771753,
        "creator_version": {
          "host_system": "Unix",
          "version": 30
        },
        "disk_nbr_start": 0,
        "external_attrs": 2175008768,
        "extra": [
          85,
          84,
          5,
          0,
          3,
          224,
          85,
          89,
          79,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "header_offset": 69,
        "internal_attrs": 1,
        "method": "Store",
        "mode": 420,
        "modified": "2012-03-09T00:59:12Z",
        "name": "bar.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uncompressed_size": 4
      },
      "local": {
        "accessed": null,
        "compressed_size": 4,
        "crc32": 77771753,
        "created": null,
        "extra": [
          85,
          84,
          9,
          0,
          3,
          224,
          85,
          89,
          79,
          224,
          85,
          89,
          79,
          117,
          120,
          11,
          0,
          1,
          4,
          245,
          1,
          0,
          0,
          4,
          20,
          0,
          0,
          0
        ],
        "flags": 0,
        "gid": 501,
        "header_offset": 0,
        "method": "Store",
        "method_specific": "None",
        "mode": 0,
        "modified": "2012-03-09T00:59:12Z",
        "name": "bar.txt",
        "reader_version": {
          "host_system": "MsDos",
          "version": 10
        },
        "uid": 501,
        "uncompressed_size": 4
      }
    }
  ],
  "encoding": "Utf8",
  "eocd": {
    "dir": {
      "inner": {
        "dir_disk_nbr": 0,
        "dir_records_this_disk": 2,
        "directory_offset": 138,
        "directory_records": 2,
        "directory_size": 154,
        "disk_nbr": 0
      },
      "offset": 292
    },
    "dir64": null,
    "global_offset": 0
  },
  "parsed_ranges": [
    {
      "contains": "end of central directory record",
      "end": 314,
      "start": 292
    },
    {
      "contains": "central directory header",
      "end": 215,
      "filename": "foo.txt",
      "start": 138
    },
    {
      "contains": "central directory header",
      "end": 292,
      "filename": "bar.txt",
      "start": 215
    },
    {
      "contains": "local file header",
      "end": 65,
      "filename": "foo.txt",
      "start": 0
    },
    {
      "contains": "file data",
      "end": 69,
      "filename": "foo.txt",
      "start": 65
    },
    {
      "contains": "local file header",
      "end": 134,
      "filename": "bar.txt",
      "start": 69
    },
    {
      "contains": "file data",
      "end": 138,
      "filename": "bar.txt",
      "start": 134
    }
  ],
  "size": 314
}
